linera-core/src/client/validator_manager/manager.rs

```diff
@@ -0,0 +1,1268 @@
+///     200,                     // cache up to 200 entries
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct ValidatorManager<Env: Environment> {
+    /// Thread-safe map of validator nodes indexed by their public keys.
+    /// Each node is wrapped with EMA-based performance tracking information.
+    nodes: Arc<tokio::sync::RwLock<BTreeMap<ValidatorPublicKey, NodeInfo<Env>>>>,
+    /// Maximum number of concurrent requests allowed per node.
+    /// Prevents overwhelming individual validators with too many parallel requests.
+    max_requests_per_node: usize,
+    /// Default scoring weights applied to new nodes.
+    default_weights: ScoringWeights,
+    /// Default EMA smoothing factor for new nodes.
+    default_alpha: f64,
+    /// Default maximum expected latency in milliseconds for score normalization.
+    default_max_expected_latency_ms: f64,
+    /// Tracks in-flight requests to deduplicate concurrent requests for the same data.
+    in_flight_tracker: InFlightTracker<RemoteNode<Env::ValidatorNode>>,
+    /// Cache of recently completed requests with their results and timestamps.
+    cache: RequestsCache<RequestKey, RequestResult>,
+}
+
+impl<Env: Environment> ValidatorManager<Env> {
+    /// Creates a new `ValidatorManager` with default configuration.
+    pub fn new(nodes: impl IntoIterator<Item = RemoteNode<Env::ValidatorNode>>) -> Self {
+        Self::with_config(
+            nodes,
+            MAX_IN_FLIGHT_REQUESTS,
+            ScoringWeights::default(),
+            0.1,
+            MAX_ACCEPTED_LATENCY_MS,
+            Duration::from_secs(CACHE_TTL_SEC), // 60 second cache TTL
+            CACHE_MAX_SIZE,                     // cache up to 100 entries
+        )
+    }
+
+    /// Creates a new `ValidatorManager` with custom configuration.
+    ///
+    /// # Arguments
+    /// - `nodes`: Initial set of validator nodes
+    /// - `max_requests_per_node`: Maximum concurrent requests per node
+    /// - `weights`: Scoring weights for performance metrics
+    /// - `alpha`: EMA smoothing factor (0 < alpha < 1)
+    /// - `max_expected_latency_ms`: Maximum expected latency for score normalization
+    /// - `cache_ttl`: Time-to-live for cached responses
+    /// - `max_cache_size`: Maximum number of entries in the cache
+    pub fn with_config(
+        nodes: impl IntoIterator<Item = RemoteNode<Env::ValidatorNode>>,
+        max_requests_per_node: usize,
+        weights: ScoringWeights,
+        alpha: f64,
+        max_expected_latency_ms: f64,
+        cache_ttl: Duration,
+        max_cache_size: usize,
+    ) -> Self {
+        assert!(alpha > 0.0 && alpha < 1.0, "Alpha must be in (0, 1) range");
+        Self {
+            nodes: Arc::new(tokio::sync::RwLock::new(
+                nodes
+                    .into_iter()
+                    .map(|node| {
+                        let public_key = node.public_key;
+                        (
+                            public_key,
```

Comment:
Nit: I think this will format more nicely?

```suggestion
                        (
                            node.public_key,
```

Alternatively I'd `let node_info = NodeInfo::with_config(…)` and then return `(public_key, node_info)`.